📜Create sound tutorial guide (#62)
KinsonDigital/VelaptorDocs: docs/tutorials/guides/sounds.mdx

---
sidebar_position: 8
---

import { Audio } from "@site/src/components/Audio";
import { TutorialLink } from "@site/src/components/TutorialLink";

# Sounds

:::note Guide Source Code
Go to the <TutorialLink projectName="Sounds"/> tutorial project to see the source code for a fully working example of this guide.
:::

Sounds are a great way to add some life to your game.  Whether it is background music or sound effects, Velaptor has you covered.
Could you ever imagine playing Mario Brothers games without that sound effect when you jump?

> Or how about all of the weapon sound effects in Metroid?  Don't forget the music in Metroid which was amazing.

Reviewer: Reword
```suggestion
Can you imagine playing Mario Brothers games without the "jump" sound effect? How about playing Metroid with no amazing music and no weapon sound effects?  Sounds are a great way to add some life to your game.  Whether it is background music or sound effects, Velaptor has you covered.

```

> Check some of these out!! For some of us, this brings back some memories. 😀

Reviewer: Reword
```suggestion
Check out some of these sounds!! They might bring back memories. 😉
```

	</td>
    <td style={{ border: 'none' }}>
		<Audio soundUrl="https://github.com/KinsonDigital/VelaptorTutorials/raw/master/Sounds/Content/Sounds/metroid-ice-beam.mp3"
			   soundName="Metroid Ice Beam"/>
	</td>
    <td style={{ border: 'none' }}>
		<Audio soundUrl="https://github.com/KinsonDigital/VelaptorTutorials/raw/master/Sounds/Content/Sounds/brinstar-music.mp3"
			   soundName="Metroid Brinstar Music"/>
	</td>
  </tr>
</table>

## <span class="color-step">Step 1: Project setup</span>

Create a basic Velaptor application.

:::note Refer To Guide
For more info, refer to the [Project Setup](./project-setup) guide.  
You only need the `OnLoad()` and `OnUpdate()` methods.
:::

## <span class="color-step">Step 2: Create class fields</span>

Create the `IAppInput<KeyboardState>`, `KeyboardState`, and `ISound?` class fields.  These will be used to get the keyboard
input, hold the previous keyboard state for each loop iteration, and hold the loaded sound.

```csharp
public class Game : Window
{
    private IAppInput<KeyboardState>? keyboard;
    private KeyboardState prevKeyState;
    private ISound? music;
	...
}
```

## <span class="color-step">Step 3: Setup window</span>

> Let's set up our window to not be so large.  Create a `Game()` constructor and add the following code.

Reviewer: Grammar - Add colon
```suggestion
Let's set up our window to not be so large.  Create a `Game()` constructor and add the following code:
```

Play it to see if you like it.  😀
<Audio soundUrl="https://github.com/KinsonDigital/VelaptorTutorials/raw/feature/add-sound-guide/Sounds/Content/Sounds/deep-consistency.ogg"/>

### <span class="color-sub-step">Step 4.2: Setup file as content</span>

We need to add the audio file to the project into a particular directory and then set the file to be copied to the
build output.  This will be the default location where Velaptor will be looking for audio content.

:::note Refer To Guide
For more info, refer to the [Adding Content](./adding-content) guide.
:::

### <span class="color-sub-step">Step 4.3: Write code to load content</span>

Now that we have the audio file added to the project, we can write some code to load the audio content.  Add the
following code to the `OnLoad()` method to load the sound and create our keyboard input object.

```csharp
    protected override void OnLoad()
    {
        this.keyboard = InputFactory.CreateKeyboard();
        this.music = ContentLoader.LoadSound("deep-consistency");

        base.OnLoad();
    }
```

:::tip Content name

> Remember, the name of the audio file without its extension is the name of the content.  The audio file added previously
> is the name of the file if you downloaded the file from the tutorial project.  If you provided an audio file,
> make sure the content name matches the file name that you decided to use.

Reviewer: Reword
```suggestion
Remember, the name of the audio file **_without its extension**_ will be the name of the content. 
Example: 'deep-consistency.ogg' will become 'deep-consistency'.
```

Reviewer: It's correct, but there are _ before and after "without its extension."  Can you please remove them?  I meant to bold and italicize it.

:::

## <span class="color-step">Step 5: Displaying sound info</span>

> So we have an audio file to play, but we need to know some information such as the current duration of the audio
> if it is playing or paused, and some basic instructions for the user on how to control the audio playback.  Since we are going
> to be using the keyboard to control the sound, it makes sense to tell the user which keyboard keys to use to stop, play, and
> other sound-related operations.

Reviewer: Reword
```suggestion
So we have an audio file to play, but we need to know some information like the current duration of the audio,
if it's playing or paused, and some basic instructions for the user on how to control the audio playback.  Since we are going
to be using the keyboard to control the sound, it makes sense to tell the user which keyboard keys to use to stop, play, and
other sound-related operations.
```

> In the spirit of keeping things simple, we will just display the information in the title bar of the window.

Reviewer: Reword
```suggestion
In the spirit of keeping things simple, we'll display the information in the title bar of the window.
```

### <span class="color-sub-step">Step 5.1: Displaying audo state</span>

> We will create a method that will be used to display the information we want in the title bar.  Create the following `UpdateTitle()`

Reviewer: Reword
```suggestion
We'll create a method that will be used to display the information we want in the title bar.  Create the following `UpdateTitle()`
```

method and add some code to get a string representing the current state of the audio file.

```csharp
private void UpdateTitle()
{
	string state;

	if (this.music.State == SoundState.Stopped)
	{
		state = "Stopped";
	}
	else
	{
		state = this.music.State == SoundState.Playing ? "Playing" : "Paused";
	}
}
```

### <span class="color-sub-step">Step 5.2: Display time and instructions</span>

> Now add the following code below the code you added previously.  This code will get the minutes and seconds of the audio
> playback in the correct 2-digit format as well as the instructions for the user on how to play, pause, fast forward,
> and rewind the audio.

Reviewer: Reword
```suggestion
Now, add the code below the code you previously added.  This code will get the minutes and seconds of the audio
playback in the correct 2-digit format as well as the instructions for the user on how to play, pause, fast forward,
and rewind the audio.
```

```csharp
private void UpdateTitle()
{
	...
	var minutes = (int)this.music.Position.Minutes;
	var seconds = (int)this.music.Position.Seconds;

	var minStr = minutes <= 9 ? $"0{minutes}" : minutes.ToString();
	var secStr = seconds <= 9 ? $"0{seconds}" : seconds.ToString();

	var playTime = $"{minStr}:{secStr}";
	const string instructions = "Instructions: Space(play/pause) Esc(restart) Left(-5s) - Right(+5s)";

	Title = $"{state} - {playTime} | {instructions}";
}
```

Great!  Now we have a method to update useful information about the audio as well as instructions for the user!!

## <span class="color-step">Step 6: Controlling the audio</span>

:::note Refer To Guide
For more info, refer to the [Keyboard Input](./keyboard-input) guide.
:::

> We can put this all together so we can control the audio playback.  But to do this we need to inspect the state of the keyboard
> to see which key is being pressed.

Reviewer: Reword
```suggestion
We can put this all together so we can control the audio playback but first, we need to inspect the state of the keyboard
to see which key is being pressed.
```

### <span class="color-sub-step">Step 6.1: Keyboard state & key pressed</span>

Add the following code to the `OnUpdate()` method that will get the current state of the

> keyboard and also check if any key has been fully pressed.

Reviewer: Reword
```suggestion
To get the current state of the keyboard and to check if any key has been fully pressed, add the code below to the `OnUpdate()` method. 
```

```csharp
protected override void OnUpdate(FrameTime frameTime)
{
	var currKeyState = this.keyboard.GetState();

	bool IsPressed(KeyCode key)
	{
		return this.prevKeyState.IsKeyDown(key) && currKeyState.IsKeyUp(key);
	}
}
```

:::tip Local Functions

> This code snippet uses a local function to check if a key is pressed.  This is a new feature in C# 7.0.  If you are not
> familiar with local functions, you can read more about them [here](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/local-functions).  

Reviewer: Reword
```suggestion
This code snippet uses a local function to check whether or not a key is pressed, which is a new feature in C# 7.0.  If you are not
familiar with local functions, you can read more about them [here](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/local-functions).  
```

```csharp
// This is a local function
bool IsPressed(KeyCode key) 
{
	return this.prevKeyState.IsKeyDown(key) && currKeyState.IsKeyUp(key);
}
```
:::

### <span class="color-sub-step">Step 6.2: Audio control with keyboard</span>

> This tutorial would be kind of useless if we did not provide any control over the audio for demo and learning purposes. 😉
> Make this happen by adding the following code after the local function in the `OnUpdate()` method.

Reviewer: Reword
```suggestion
This tutorial would be pretty useless if we didn't provide control over the audio for demo and learning purposes, right?😉
You can make that happen by adding the following code after the local function in the `OnUpdate()` method.
```

		}
		else
		{
			this.music.Pause();
		}
	}
	else if (IsPressed(KeyCode.Escape))
	{
		this.music.Reset();
	}
	else if (IsPressed(KeyCode.Left))
	{
		this.music.Rewind(5);
	}
	else if (IsPressed(KeyCode.Right))
	{
		this.music.FastForward(5);
	}
}
```

### <span class="color-sub-step">Step 6.3: Finishing up</span>

Now, to finish things up, we can finalize our code by implementing keyboard state tracking for the previous frame, alongside
updating the window title. The significance of the previous keyboard state is important. Without it, we cannot accurately
determine if a key has been pressed. To accomplish this, we must diligently maintain a record of the keyboard's state in
the previous frame and subsequently compare it with the current state. This approach enables us to discern whether a key
has moved from a pressed (down) position to a released (up) position. 

To implement this, simply insert the following code snippet after the `OnUpdate()` method.

```csharp
protected override void OnUpdate(FrameTime frameTime)
{
	...

	this.prevKeyState = currKeyState;

	UpdateTitle();

	base.OnUpdate(frameTime);
}
```

## <span class="color-step">Step 7: Run it</span>

> Run the game and let's play some music!  You should see the following window.

Reviewer: Grammar - add colon
```suggestion
Run the game and let's play some music!  You should see the following window:
```

<img style={{width: '75%'}} src={require('@site/static/img/tutorials/guides/sounds/sounds-window.png').default} alt="Example banner"/>

> As the instructions in the title bar state, press the space bar to play and pause the audio, escape to stop the audio, and set
> it back to the beginning, the left arrow key to rewind 5 seconds, and the right arrow key to fast forward 5 seconds.

Reviewer: Reword
```suggestion
As the instructions in the title bar state, press the space bar to play and pause the audio, escape to stop or restart the audio, the left arrow key to rewind 5 seconds, and the right arrow key to fast forward 5 seconds.
```
